modules/shared/src/main/scala/org/tessellation/currency/dataApplication/package.scala

```diff
@@ -1,33 +1,150 @@
+import cats.effect.Async
 import cats.kernel.Eq
 import cats.syntax.all._
-import cats.{Applicative, Monad, MonadThrow}
 
 import scala.reflect.ClassTag
 import scala.util.control.NoStackTrace
 
+import org.tessellation.currency.dataApplication.DataTransaction.DataTransactions
+import org.tessellation.currency.dataApplication.FeeTransaction.{decoderFeeTransactionRequest, serialize}
 import org.tessellation.currency.dataApplication.dataApplication.{DataApplicationBlock, DataApplicationValidationErrorOr}
 import org.tessellation.currency.schema.EstimatedFee
 import org.tessellation.currency.schema.currency.{CurrencyIncrementalSnapshot, CurrencySnapshotInfo}
-import org.tessellation.currency.schema.feeTransaction.FeeTransaction
+import org.tessellation.ext.derevo.ordering
+import org.tessellation.json.JsonBinarySerializer
 import org.tessellation.routes.internal.ExternalUrlPrefix
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Amount
 import org.tessellation.schema.round.RoundId
 import org.tessellation.schema.{GlobalIncrementalSnapshot, SnapshotOrdinal}
+import org.tessellation.security._
 import org.tessellation.security.hash.Hash
 import org.tessellation.security.signature.Signed
-import org.tessellation.security.{Encodable, Hashed, SecurityProvider}
 
+import derevo.cats.{order, show}
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
 import eu.timepit.refined.auto._
 import io.circe._
 import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
 import io.circe.syntax._
 import org.http4s._
-import org.http4s.circe.jsonEncoderOf
+import org.http4s.circe.{jsonEncoderOf, jsonOf}
 import org.http4s.server.Router
 
-trait DataUpdate
+sealed trait DataTransaction
+object DataTransaction {
+  type DataTransactions = NonEmptyList[Signed[DataTransaction]]
+
+  implicit def encoder(implicit e: Encoder[DataUpdate]): Encoder[DataTransaction] =
+    Encoder.instance {
+      case v: FeeTransaction => v.asJson
+      case v: DataUpdate     => e.apply(v)
+    }
+
+  implicit def decoder(implicit d: Decoder[DataUpdate]): Decoder[DataTransaction] =
+    (c: HCursor) => Decoder[FeeTransaction].apply(c).orElse(d.apply(c))
+
+  private def collectTransactions[A <: DataTransaction](dataTransactions: List[DataTransactions])(
+    pf: PartialFunction[DataTransaction, A]
+  ): List[Signed[A]] =
+    dataTransactions.flatMap { currentDataTransactions =>
+      currentDataTransactions.collect { dataTransaction =>
+        dataTransaction.value match {
+          case transaction if pf.isDefinedAt(transaction) =>
+            Signed(pf(transaction), dataTransaction.proofs)
+        }
+      }
+    }
+
+  def getDataUpdates(dataTransactions: List[DataTransactions]): List[Signed[DataUpdate]] =
+    collectTransactions(dataTransactions) {
+      case dataUpdate: DataUpdate => dataUpdate
+    }
+
+  def getFeeTransactions(dataTransactions: List[DataTransactions]): List[Signed[FeeTransaction]] =
```

Comment:
ìt should be defined in FeeTransaction companion object

Reply:
👍🏼